lgc/patch/PatchBoundsCheckMemory.cpp

```diff
@@ -0,0 +1,195 @@
+}
+
+static Value *getTypeZero(Type *ty) {
+  if (ty->isIntegerTy()) {
+    return ConstantInt::get(ty, 0);
+  } else if (ty->isFloatTy() || ty->isHalfTy() || ty->isDoubleTy() || ty->isFP128Ty() || ty->isBFloatTy()) {
+    return ConstantFP::get(ty, 0.0);
+  } else {
+    return ConstantAggregateZero::get(ty);
+  }
+}
+
+// =====================================================================================================================
+PatchBoundsCheckMemory::PatchBoundsCheckMemory() : Patch(ID) {
+}
+
+// =====================================================================================================================
+// Executes this pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+bool PatchBoundsCheckMemory::runOnModule(Module &module) {
+  LLVM_DEBUG(dbgs() << "Run the pass llpc-bounds-check-memory\n");
+
+  m_pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(&module);
+
+  if (!m_pipelineState->getOptions().enableScratchBoundsCheck)
+    return false;
+
+  m_context = &module.getContext();
+  m_builder.reset(new IRBuilder<>(*m_context));
+
+  visit(module);
+
+  bool changed = false;
+  for (auto &inst : m_getElementPtrInsts)
+    changed |= addBoundsCheck(inst);
+  m_getElementPtrInsts.clear();
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Visits "getelementptr" instruction.
+//
+// @param getElemPtr : "GetElementPtr" instruction
+void PatchBoundsCheckMemory::visitGetElementPtrInst(GetElementPtrInst &getElemPtr) {
+  auto pointerTy = getElemPtr.getPointerOperand()->getType();
+  if (pointerTy->getPointerAddressSpace() != ADDR_SPACE_PRIVATE)
+    return;
+
+  // Check for dynamic indices in the instruction
+  llvm::SmallVector<std::pair<Value *, uint64_t>, 1> dynIndices;
+  std::vector<Value *> indices;
+  for (unsigned i = 1, operandCount = getElemPtr.getNumOperands(); i < operandCount; ++i) {
+    auto index = getElemPtr.getOperand(i);
+    if (!isa<Constant>(index)) {
+      auto indexedTy = GetElementPtrInst::getIndexedType(pointerTy->getPointerElementType(), indices);
+      if (auto vectorTy = dyn_cast<FixedVectorType>(indexedTy)) {
+        dynIndices.push_back({index, vectorTy->getNumElements()});
+      } else if (auto arrayTy = dyn_cast<ArrayType>(indexedTy)) {
+        dynIndices.push_back({index, arrayTy->getNumElements()});
+      } else {
+        llvm_unreachable("Unsupported indexed type for bounds checking");
+      }
+      break;
+    }
+    indices.push_back(index);
+  }
+  if (dynIndices.empty())
+    return;
+  m_getElementPtrInsts.push_back({&getElemPtr, dynIndices});
+}
+
+// =====================================================================================================================
+// Checks whether each dynamic index in the specified "getelementptr" instruction is lower than its allowed upper bound.
```

Comment:
What if the index is just one past the size of the upper bound but the pointer is never accessed?

Reply:
We only *change* the behavior of load and store instructions. If there are no loads or stores, the `inBounds` condition might be false but nothing changes.